### ISSUE TITLE
Explore Classmethods for alternative constructors

### ISSUE THREAD

**Issue by ljwolf** [pysal/pysal#814]
In PySAL, we use quite a few of these kinds of patterns:
```
pysal.queen_from_shapefile
pysal.rook_from_shapefile
```
as well as functions like:
```
ps.weights.util.WSP2W
ps.weights.util.W2WSP
```
It might be helpful to explore making these classmethods. This would look similar to how pandas's `DataFrame` constructor has a few alternative constructors:
```
pd.DataFrame.from_csv()
pd.DataFrame.from_dict()
pd.DataFrame.from_items()
pd.DataFrame.from_records()
```
Formula-based or dataframe-based operations then could be implemented as classmethods:
```
ps.spreg.ML_Lag.from_formula('HOVAL ~ CRIME + INC', data=df)
ps.W.from_shapefile('NAT.shp', wtype='queen')
ps.Adaptive_Bandwidth.from_dataframe(df, geometry='the_geom', k=4)
ps.W.from_WSP(W, silent_island_warning=True)
```
This would have a few pretty desirable benefits:

1. This follows "explicit is better than implicit" while having "one and only one obvious way to do things." A polymorphic/preparser strategy gives one obvious way to do things, but does all of the conversion & parsing implicitly. 
2. Visually, we can retain the `*_from_shapefile` appearance while centralizing constructors. All of the classmethods would target the class's default init, which would be the "lowest level" constructor for the application. 
3. We can add new optional interfaces that exploit all the fancy stuff we like: classmethods can can be guarded by [@requires](https://github.com/pysal/pysal/blob/master/pysal/spreg/optional_imports.ipynb). would look something like this:
```python
class ML_Lag(object):
    ...
    @classmethod
    @requires('patsy', 'pandas')
    def from_formula(formulalike, df, *args, **kwargs):
        y,X = parse_formula_using_patsy(formulalike, df)
        self.__init__(y, X, *args, **kwargs)
```

**Comment by jlaura** [comment]
This is the canonical example of using classmethods in Python.  Definitely a big 👍here.  Is pandas using the classmethod idiom, subclassing from a baseclass that has these methods and dispatching or a combination of the two?

**Comment by ljwolf** [comment]
as far as I can tell, they're mostly defined in the [`DataFrame` subclass of an ABC](https://github.com/pydata/pandas/blob/master/pandas/core/frame.py#L777). 

We could still support interception/preparsing so that we have a "smart dispatcher," but that'd essentially typecheck & send to the correct classmethod, forcing either an init or a `NotImplemented`/`ImportError` if needed.

**Comment by sjsrey** [comment]
This looks great.

I'm wondering, as this relates to the [wconstructor thread](https://github.com/pysal/pysal/wiki/AbstractionWeights), about the following which is supported:

```
wq = ps.W.from_shapefile('NAT.shp', wtype='queen')
wr = ps.W.from_shapefile('NAT.shp', wtype='rook')
```

versus the alternative use case of something like:

```
collection = ps.PolygonCollection_from_shapefile('NAT.shp')
wq = ps.contiguity_from_collection(collection, wtype='queen')
wr = ps.contiguity_from_collection(collection, wtype='rook')
```

The second case is only 1 file read and keeps the collection around if that were needed for other things besides alternative weights.

**Comment by ljwolf** [comment]
Yep, that's definitely a target. I was thinking three classmethods, in addition to the init. 
```
from_iterable
from_dataframe
from_shapefile
```
the `__init__` would take whatever the most basic `read()` command targets. Right now, this is a list of shapes.
`from_iterable` would take something that supports iteration and cast to what `__init__` needs
`from_dataframe` would extract ids/geometries from the dataframe and treat them like an iterable.
`from_shapefile` would do the same for shapefiles. 

This should be currently supported by the feature branch:
```python
f = ps.open('NAT.shp')
collection = f.read()
wq = ps.Queen(collection)
wr = ps.Rook(collection)
```

**Comment by ljwolf** [comment]
Oh yes, and a `to_WSP`/`from_WSP`, and matching methods on `WSP.

**Comment by sjsrey** [comment]
What I'm not clear on is where you are pulling in the PolygonCollection class which is needed by ContiguityWeightsPolygons  but I don't see PolygonCollection in cg/shapes.py in this branch?

**Comment by ljwolf** [comment]
Classmethods for this in weights are implemented over in my [GSOC feature branch](https://github.com/ljwolf/pysal/tree/pdio). ([example notebook](https://github.com/ljwolf/pysal/blob/pdio/from%20classmethods.ipynb))

One remaining question: I'm encouraging the type of interaction we see in stuff like `rook_from_shapefile` becoming `Rook.from_shapefile`, rather than `W.from_shapefile(...,type='rook')`.

Supporting this kind of dispatch is doable, just need to build a dispatch table. Is this what we want? 

In addition, should I remap the functions in `user.py` to call the classmethods, or just leave them alone?

**Comment by ljwolf** [comment]
Implemented in #821.
